File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
  public List<Integer> findAnagrams(String s, String p) {
    List<Integer> result = new ArrayList();
    int[] sCount = new int[26];
    int[] pCount = new int[26];
    for(char c: p.toCharArray()) {
        pCount[c-'a']++;
    }
    int left = 0, k = p.length();
    for(int right = 0; right < s.length(); right++) {
        char l = s.charAt(left);
        char r = s.charAt(right);

        sCount[r-'a']++;

        if(right - left + 1 == k) {
            if(Arrays.equals(pCount,sCount)){
                result.add(left);
            }
            sCount[l-'a']--;
            left++;
        }
    }

    return result;
  }

}

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
from collections import Counter

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        result = []
        pCount = Counter(p)
        sCount = Counter()
        left = 0;
        k = len(p)
        
        for right in range(0,len(s)):
            sCount[s[right]] += 1
            if(right - left + 1 == k):

                if pCount == sCount:
                    result.append(left)

                if sCount[s[left]] == 1:
                    del sCount[s[left]]
                else:
                    sCount[s[left]] -= 1

                left += 1
        return result
